src/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,433 @@
+
+This API is about exposing the `ReferenceHandling` property as a class type, in order to make it very extensible for future; and provide built-in static instances of `Default` and `Preserve` that are useful to enable the most common behaviors by just setting those in `JsonSerializerOptions.ReferenceHandling`.
+
+With ReferenceHandling being a class, we can exclude things that, as of now, we are not sure if they need to be shipped, in example, the `ReferenceLoopHandling.Ignore` option or the Object and Array granularity of Json.Net's `PreserveReferencesHandling`.
+
+In a future, we might also be able to extend the Deserialization behaviors, in order to provide the granularity of Serialization on Deserialization. Or we could also be able to define independant behaviors for both Serialization and Deserialization; i.e: someone may want to emit payloads with preserved references but he does not want to read them.
+
+# Examples
+
+Having the following class:
+```cs
+class Employee 
+{ 
+    string Name { get; set; }
+    Employee Manager { get; set; }
+    List<Employee> Subordinates { get; set; }
+}
+```
+
+## Using Preserve on Serialize
+On System.Text.Json:
+```cs
+public static void WritePreservingReference()
+{
+    var bob = new Employee { Name = "Bob" };
+    var angela = new Employee { Name = "Angela" };
+
+    angela.Manager = bob;
+    bob.Subordinates = new List<Employee>{ angela };
+
+    var options = new JsonSerializerOptions
+    {
+        ReferenceHandling = ReferenceHandling.Preserve
+        WriteIndented = true,
+    };
+
+    string json = JsonSerializer.Serialize(angela, options);
+    Console.Write(json);
+}
+```
+
+On Newtonsoft's Json.Net:
+```cs
+public static void WritePreservingReference()
+{
+    var bob = new Employee { Name = "Bob" };
+    var angela = new Employee { Name = "Angela" };
+
+    angela.Manager = bob;
+    bob.Subordinates = new List<Employee>{ angela };
+
+    var settings = new JsonSerializerSettings
+    {
+        PreserveReferencesHandling = PreserveReferencesHandling.All
+        Formatting = Formatting.Indented 
+    };
+
+    string json = JsonConvert.SerializeObject(angela, settings);
+    Console.Write(json);
+}
+```
+
+Output:
+```jsonc
+{
+    "$id": "1",
+    "Name": "Angela",
+    "Manager": {
+        "$id": "2",
+        "Name": "Bob",
+        "Subordinates": { //Note how the Subordinates' square braces are replaced with curly braces in order to include $id and $values properties, $values will now hold whatever value was meant for the Subordinates list.
+            "$id": "3",
+            "$values": [
+                {  //Note how this object denotes reference to Angela that was previously serialized.
+                    "$ref": "1"
+                }
+            ]
+        }            
+    }
+}
+```
+
+## Using Preserve on Deserialize
```

Comment:
Show sample example of what the usage might look like if we add the APIs in vNext (outside of the built-in factory methods).